### PR TITLE
Implement model view model and dal for edit patient

### PR DESCRIPTION
**Pull request by kmitche9**

### CODE REVIEW

File: CS3230Project/View/EditPatient.cs [hunk 1]
```diff
@@ -10,38 +11,71 @@ namespace CS3230Project.View
...
         {
             this.InitializeComponent();
+            this.updatedDetails = new Dictionary<string, string>();
             this.bindLabelsToCurrentUser();
             this.setStatusComboBoxValues();
+            this.setGenderComboBoxValues();
             this.loadPatientData(patient);
+            this.updatedDetails.Add("PatientId", patient.PatientId.ToString());
+
+            this.states = new[]
```
Comment (alexofceaser85) [inline CS3230Project/View/EditPatient.cs]: The dropdowns should have an items property where you can input the states. For some reason the items are not showing for the register patient on your branch but they are showing on master. It should not be an issue but just be aware when merging the branches.

Capture

File: CS3230Project/Model/Users/Patients/Patient.cs [hunk 2]
```diff
@@ -107,101 +45,107 @@ public class Patient
...
-            {
-                throw new ArgumentException(PatientErrorMessages.PhoneNumberCannotBeNull);
-            }
+            if (city == null) throw new ArgumentException(PatientErrorMessages.CityCannotBeNull);
+            if (city.Trim().Length == 0) throw new ArgumentException(PatientErrorMessages.CityCannotBeEmpty);
+            if (state == null) throw new ArgumentException(PatientErrorMessages.StateCannotBeNull);
+            if (state.Trim().Length == 0) throw new ArgumentException(PatientErrorMessages.StateCannotBeEmpty);
+            if (zipcode == null) throw new ArgumentException(PatientErrorMessages.ZipcodeCannotBeNull);
+            if (zipcode.Trim().Length == 0) throw new ArgumentException(PatientErrorMessages.ZipcodeCannotBeEmpty);
+            if (phoneNumber == null) throw new ArgumentException(PatientErrorMessages.PhoneNumberCannotBeNull);
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/Patient.cs]: This looks real messy, I don't know why you removed the braces but they need to be put back in.

File: CS3230Project/Model/Users/Patients/Patient.cs [hunk 3]
```diff
@@ -4,75 +4,13 @@
...
-        /// <summary>
-        /// The phone number for the patient
-        /// </summary>
-        public string PhoneNumber { get; }
-
-        /// <summary>
-        /// The status status for the patient
-        /// </summary>
-        public bool Status { get; }
-
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/Patient.cs:71]: Why did you move the properties from the top to the bottom?

File: CS3230Project/Model/Users/Patients/PatientManager.cs [hunk 4]
```diff
@@ -49,23 +46,19 @@ public static List<Patient> GetPatientsByName(string firstName, string lastName)
...
         {
             if (dateOfBirth < new DateTime(1900, 1, 1))
-            {
                 throw new ArgumentException(PatientErrorMessages.DateOfBirthCannotBeBefore1900);
-            }
 
             if (dateOfBirth > DateTime.Now)
-            {
                 throw new ArgumentException(PatientErrorMessages.DateOfBirthCannotBeInTheFuture);
-            }
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs]: Why did you remove the braces?

File: CS3230Project/Model/Users/Patients/PatientManager.cs [hunk 5]
```diff
@@ -28,16 +27,14 @@ public static List<Patient> GetPatientsByName(string firstName, string lastName)
         {
             if ((firstName == null || firstName.Trim().Length == 0) &&
                 (lastName == null || lastName.Trim().Length == 0))
-            {
                 throw new ArgumentException(PatientErrorMessages.FirstAndLastNamesCannotBothBeEmpty);
-            }
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs]: Another place where the braces are removed

File: CS3230Project/Model/Users/Patients/PatientManager.cs [hunk 6]
```diff
@@ -82,20 +75,33 @@ public static List<Patient> GetPatientsByDateOfBirth(DateTime dateOfBirth)
...
             if (dateOfBirth < new DateTime(1900, 1, 1))
-            {
                 throw new ArgumentException(PatientErrorMessages.DateOfBirthCannotBeBefore1900);
-            }
             if (dateOfBirth > DateTime.Now)
-            {
                 throw new ArgumentException(PatientErrorMessages.DateOfBirthCannotBeInTheFuture);
-            }
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs]: Another place the braces are removed

File: CS3230Project/Model/Users/Patients/PatientManager.cs [hunk 7]
```diff
@@ -82,20 +75,33 @@ public static List<Patient> GetPatientsByDateOfBirth(DateTime dateOfBirth)
...
+        ///     Modifies the patient.
+        ///     Precondition:
+        ///     updatedDetails != null AND updatedDetails.Count > 0
+        /// </summary>
+        /// <param name="updatedDetails">The updated details.</param>
+        /// <returns>
+        ///     True, if the patient was modified
+        /// </returns>
+        /// <exception cref="System.ArgumentException"></exception>
+        public static bool ModifyPatient(Dictionary<string, string> updatedDetails)
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs:122]: I'm not a fan of having a dictionary, you pretty much have to remember all the fields of patient to access the dictionary fields. I would rather just pass all the fields separately or pass in a new Patient object, it is all being parsed into json into the database anyways. You also need validation for modifying a patient. The same validation as adding a new patient.

File: CS3230Project/Model/Users/Patients/PatientManager.cs [hunk 8]
```diff
@@ -82,20 +75,33 @@ public static List<Patient> GetPatientsByDateOfBirth(DateTime dateOfBirth)
...
+        {
+            if (updatedDetails == null)
+                throw new ArgumentException(PatientErrorMessages.UpdatedPatientDetailsCannotBeNull);
+            if (updatedDetails.Count == 0)
+                throw new ArgumentException(PatientErrorMessages.UpdatedPatientDetailsCannotBeEmpty);
+
```
Comment (alexofceaser85) [inline CS3230Project/Model/Users/Patients/PatientManager.cs:132]: And you need braces.

File: CS3230Project/DAL/Patients/PatientsDal.cs [hunk 9]
```diff
@@ -209,10 +205,102 @@ private static List<Patient> createPatients(MySqlCommand command, string firstNa
...
+            var numberOfDetailsChanged = 0;
+            var patientId = "";
+
+            foreach (var currDetail in updatedDetails)
+            {
+                if (numberOfDetailsChanged > 0 && !commandText[commandText.Length - 2].Equals(',') &&
+                    !currDetail.Key.Equals("PatientId")) commandText += ", ";
+
+                switch (currDetail.Key)
+                {
```
Comment (alexofceaser85) [inline CS3230Project/DAL/Patients/PatientsDal.cs:246]: At some point we need to look into refactoring this. It probably will just be cleaner to update all the fields. We will have to take another look at this method after this iteration.

File: CS3230Project/DAL/Patients/PatientsDal.cs [hunk 10]
```diff
@@ -209,10 +205,102 @@ private static List<Patient> createPatients(MySqlCommand command, string firstNa
                     reader.GetFieldValueCheckNull<string>(zipcodeOrdinal),
                     reader.GetFieldValueCheckNull<string>(phoneNumberOrdinal),
                     reader.GetFieldValue<bool>(statusOrdinal)));
-            }
```
Comment (alexofceaser85) [inline CS3230Project/DAL/Patients/PatientsDal.cs]: Needs braces

**Review by alexofceaser85 — CHANGES_REQUESTED** [review]

File: CS3230Project/View/EditPatient.cs (hunk 1 above)
Comment (kmitche9) [reply CS3230Project/View/EditPatient.cs]: I have removed from the code, and added it through the editor. I will check the Register patient once merged.

File: CS3230Project/Model/Users/Patients/Patient.cs (hunk 3 above)
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/Patient.cs:71]: Resharper... Has been fixed.

File: CS3230Project/Model/Users/Patients/Patient.cs (hunk 2 above)
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/Patient.cs]: Again, Resharper... Has been fixed.

File: CS3230Project/Model/Users/Patients/PatientManager.cs (hunk 5 above)
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs]: Fixed.

File: CS3230Project/Model/Users/Patients/PatientManager.cs (hunk 4 above)
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs]: fixed.

File: CS3230Project/Model/Users/Patients/PatientManager.cs (hunk 6 above)
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs]: fixed

File: CS3230Project/Model/Users/Patients/PatientManager.cs (hunk 8 above)
Comment (kmitche9) [reply CS3230Project/Model/Users/Patients/PatientManager.cs:132]: fixed

File: CS3230Project/DAL/Patients/PatientsDal.cs (hunk 10 above)
Comment (kmitche9) [reply CS3230Project/DAL/Patients/PatientsDal.cs]: fixed.

File: CS3230Project/DAL/Patients/PatientsDal.cs (hunk 9 above)
Comment (kmitche9) [reply CS3230Project/DAL/Patients/PatientsDal.cs:246]: I agree! I will put some thought into it.

**Review by alexofceaser85 — APPROVED** [review]
Looks good
